blockchain/types/transaction.go

```diff
@@ -308,15 +308,15 @@ func (tx *Transaction) EffectiveGasTip(baseFee *big.Int) *big.Int {
 	return tx.GasPrice()
 }
 
-func (tx *Transaction) EffectiveGasPrice(header *Header) *big.Int {
+func (tx *Transaction) EffectiveGasPrice(header *Header, config *params.ChainConfig) *big.Int {
```

Comment:
Can we clarify?

- The EffectiveGasTip should return 0 before Kaia. This function hasn't been used before Kaia so it was okay to return nonzero, but since we are going to use it we need to fix it.
- The other tx types do have nonzero tip and must be considered in EffectiveGasPrice.

```go
// Note that this function is 
func (tx *Transaction) EffectiveGasTip(header *Header, config *params.ChainConfig) *big.Int {
	var baseFee *big.Int = nil
	if header != nil && header.BaseFee != nil {
		baseFee = header.BaseFee
	}

	var tip *big.Int
	if tx.Type == 2 {
		tip = te.GetGasTipCap()
	} else {
		tip = tx.GasPrice()
	}

	if IsKaia && baseFee != nil { // After Kaia
		return Min(tip, FeeCap - baseFee)
	}
	// No tip before Kaia
	return 0
}

func (tx *Transaction) EffectiveGasPrice(header *Header, config *params.ChainConfig) *big.Int {
	var baseFee *big.Int = nil
	if header != nil && header.BaseFee != nil {
		baseFee = header.BaseFee
	}

	if IsKaia && baseFee != nil { // After Kaia
		tip := tx.EffectiveGasTip(header, config)
		return tip + BaseFee
	}
	if baseFee != nil { // After Magma
		return baseFee
	}
	// Before Magma, return (other types) GasPrice or (type 2) FeeCap
	return tx.GasPrice() 
```

Reply:
Will be handled in follow-up PR.